### PR TITLE
[SC-583] Optimize AddressArray and AddressSet libs

### PR DESCRIPTION
**Pull request by k06a**
```
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  get     ·  23901  ·  28945  ·  25539  ·  10  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  length  ·      -  ·      -  ·  23359  ·   2  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  pop     ·  21683  ·  26767  ·  24225  ·   4  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  push    ·  44145  ·  49192  ·  46307  ·  28  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  set     ·  27166  ·  29253  ·  27866  ·   3  │
·····················|··········|·········|·········|·········|·······
|  AddressSetMock    ·  add     ·  66455  ·  71502  ·  67939  ·  17  │
·····················|··········|·········|·········|·········|·······
|  AddressSetMock    ·  remove  ·  25948  ·  34664  ·  30306  ·   2  │
·····················|··········|·········|·········|·········|·······
```
=>
```
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  get     ·  23730  ·  28353  ·  25162  ·  10  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  length  ·      -  ·      -  ·  23347  ·   2  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  pop     ·  26488  ·  28510  ·  27499  ·   4  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  push    ·  43993  ·  49015  ·  46144  ·  28  │
·····················|··········|·········|·········|·········|·······
|  AddressArrayMock  ·  set     ·  27055  ·  29171  ·  27764  ·   3  │
·····················|··········|·········|·········|·········|·······
|  AddressSetMock    ·  add     ·  66295  ·  71317  ·  67772  ·  17  │
·····················|··········|·········|·········|·········|·······
|  AddressSetMock    ·  remove  ·  27525  ·  29547  ·  28536  ·   2  │
·····················|··········|·········|·········|·········|·······
```

### CODE REVIEW

**Comment by zZoMROT** [comment]
```
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  get        ·      23825  ·      28743  ·      25394  ·  10  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  length     ·          -  ·          -  ·      23404  ·   2  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  pop        ·      21677  ·      26747  ·      24212  ·   4  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  push       ·      43986  ·      48980  ·      46125  ·  28  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  set        ·      26903  ·      29011  ·      27610  ·   3  ·
·····················|·············|·············|·············|·············|······|
|  AddressSetMock    ·  add        ·      66097  ·      71137  ·      67579  ·  17  ·
·····················|·············|·············|·············|·············|······|
|  AddressSetMock    ·  remove     ·      25819  ·      34322  ·      30071  ·   2  ·
·····················|·············|·············|·············|·············|······|
```
=>
```
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  get        ·      23590  ·      28211  ·      25024  ·  10  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  length     ·          -  ·          -  ·      23329  ·   2  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  pop        ·      26371  ·      26382  ·      26377  ·   4  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  push       ·      43864  ·      48862  ·      46005  ·  28  ·
·····················|·············|·············|·············|·············|······|
|  AddressArrayMock  ·  set        ·      26860  ·      28957  ·      27563  ·   3  ·
·····················|·············|·············|·············|·············|······|
|  AddressSetMock    ·  add        ·      66046  ·      71044  ·      67516  ·  17  ·
·····················|·············|·············|·············|·············|······|
|  AddressSetMock    ·  remove     ·      32148  ·      34243  ·      33196  ·   2  ·
·····················|·············|·············|·············|·············|······|
```

**Comment by k06a** [comment]
@zZoMROT pop() didn't receive 5k gas refund, but optimizes future push() by 15k.

**Comment by ZumZoom** [comment]
```
·····················|·························|·············|·············|·················|···············|
|  Contract          ·  Method                 ·  Min        ·  Max        ·  Avg            ·  # calls      ·
·····················|·························|·············|·············|·················|···············|
|  AddressArrayMock  ·  pop                    ·      21677  ·      26747  ·          24212  ·            4  ·
·····················|·························|·············|·············|·················|···············|
|  AddressArrayMock  ·  push                   ·      43986  ·      48980  ·          46125  ·           28  ·
·····················|·························|·············|·············|·················|···············|
|  AddressArrayMock  ·  set                    ·      26903  ·      29011  ·          27610  ·            3  ·
·····················|·························|·············|·············|·················|···············|
|  AddressSetMock    ·  add                    ·      66097  ·      71137  ·          67579  ·           17  ·
·····················|·························|·············|·············|·················|···············|
|  AddressSetMock    ·  remove                 ·      25819  ·      34322  ·          30071  ·            2  ·
·····················|·························|·············|·············|·················|···············|
|  TokenWithBySig    ·  bySig                  ·      86539  ·     144238  ·         113485  ·            8  ·
·····················|·························|·············|·············|·················|···············|
```

=>

```
·····················|·························|·············|·············|·················|···············|
|  Contract          ·  Method                 ·  Min        ·  Max        ·  Avg            ·  # calls      ·
·····················|·························|·············|·············|·················|···············|
|  AddressArrayMock  ·  pop                    ·          -  ·          -  ·          26363  ·            4  ·
·····················|·························|·············|·············|·················|···············|
|  AddressArrayMock  ·  push                   ·      43864  ·      48862  ·          46005  ·           28  ·
·····················|·························|·············|·············|·················|···············|
|  AddressArrayMock  ·  set                    ·      26832  ·      28942  ·          27539  ·            3  ·
·····················|·························|·············|·············|·················|···············|
|  AddressSetMock    ·  add                    ·      66046  ·      71044  ·          67516  ·           17  ·
·····················|·························|·············|·············|·················|···············|
|  AddressSetMock    ·  remove                 ·      32021  ·      39502  ·          35762  ·            2  ·
·····················|·························|·············|·············|·················|···············|
|  TokenWithBySig    ·  bySig                  ·     106120  ·     179414  ·         139979  ·            8  ·
·····················|·························|·············|·············|·················|···············|
```

**Review by ZumZoom — DISMISSED** [review]

**Comment by ZumZoom** [comment]
Multi push/pop tests that show how dirty storage affects add/push:

```
·····················|··········|·············|·············|·················|
|  AddressSetMock    ·  add     ·      66097  ·      71137  ·          68617  ·
·····················|··········|·············|·············|·················|
|  AddressSetMock    ·  remove  ·      25819  ·      29883  ·          27851  ·
·····················|··········|·············|·············|·················|
|  AddressArrayMock  ·  pop     ·      21677  ·      26747  ·          24212  ·
·····················|··········|·············|·············|·················|
|  AddressArrayMock  ·  push    ·      43986  ·      48980  ·          46483  ·
·····················|··········|·············|·············|·················|
```

=>


```
·····················|··········|·············|·············|·················|
|  AddressSetMock    ·  add     ·      31874  ·      71044  ·          50759  ·
·····················|··········|·············|·············|·················|
|  AddressSetMock    ·  remove  ·      32021  ·      34116  ·          33069  ·
·····················|··········|·············|·············|·················|
|  AddressArrayMock  ·  pop     ·          -  ·          -  ·          26363  ·
·····················|··········|·············|·············|·················|
|  AddressArrayMock  ·  push    ·      26764  ·      48862  ·          37113  ·
·····················|··········|·············|·············|·················|
```

tldr:

```
AddressSet.add     -34K
AddressSet.remove  +4.5K
AddressArray.push  -17K
AddressArray.pop   +4.5K
```

**Review by ZumZoom — APPROVED** [review]
